lib/runtime/offchain/httpset.go

```diff
@@ -0,0 +1,103 @@
+package offchain
+
+import (
+	"errors"
+	"net/http"
+	"sync"
+)
+
+const maxConcurrentRequests = 1000
+
+var (
+	errIntBufferEmpty        = errors.New("int buffer exhausted")
+	errIntBufferFull         = errors.New("int buffer is full")
+	errRequestIDNotAvailable = errors.New("request id not available")
+)
+
+// requestIDBuffer created to control the amount of available non-duplicated ids
+type requestIDBuffer chan int16
+
+// newIntBuffer creates the request id buffer starting from 1 till @buffSize (by default @buffSize is 1000)
+func newIntBuffer(buffSize int16) *requestIDBuffer {
+	b := make(chan int16, buffSize)
+	for i := int16(1); i <= buffSize; i++ {
+		b <- i
+	}
+
+	intb := requestIDBuffer(b)
+	return &intb
+}
+
+func (b *requestIDBuffer) get() (int16, error) {
+	select {
+	case v := <-*b:
+		return v, nil
+	default:
+		return 0, errIntBufferEmpty
+	}
+}
+
+func (b *requestIDBuffer) put(i int16) error {
+	select {
+	case *b <- i:
+		return nil
+	default:
+		return errIntBufferFull
+	}
+}
+
+// HTTPSet holds a pool of concurrent http request calls
+type HTTPSet struct {
+	mtx    *sync.Mutex
+	reqs   map[int16]*http.Request
```

Comment:
is there a chance this map will be accesses concurrently? I think potentially

Reply:
yes! There is mutexes locks in `StartRequest` and `Remove` and I've added locks on `Get` tho